src/main/java/org/opensearch/geospatial/ip2geo/action/UpdateDatasourceRequest.java

```diff
@@ -0,0 +1,168 @@
+        PARSER.declareLong((request, val) -> request.setUpdateInterval(TimeValue.timeValueDays(val)), UPDATE_INTERVAL_IN_DAYS_FIELD);
+    }
+
+    /**
+     * Constructor
+     * @param name name of a datasource
+     */
+    public UpdateDatasourceRequest(final String name) {
+        this.name = name;
+    }
+
+    /**
+     * Constructor
+     * @param in the stream input
+     * @throws IOException IOException
+     */
+    public UpdateDatasourceRequest(final StreamInput in) throws IOException {
+        super(in);
+        this.name = in.readString();
+        this.endpoint = in.readOptionalString();
+        this.updateInterval = in.readOptionalTimeValue();
+    }
+
+    @Override
+    public void writeTo(final StreamOutput out) throws IOException {
+        super.writeTo(out);
+        out.writeString(name);
+        out.writeOptionalString(endpoint);
+        out.writeOptionalTimeValue(updateInterval);
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
+        ActionRequestValidationException errors = new ActionRequestValidationException();
+        if (endpoint == null && updateInterval == null) {
+            errors.addValidationError("no values to update");
+        }
+
+        validateEndpoint(errors);
+        validateUpdateInterval(errors);
+
+        return errors.validationErrors().isEmpty() ? null : errors;
+    }
+
+    /**
+     * Conduct following validation on endpoint
+     * 1. endpoint format complies with RFC-2396
+     * 2. validate manifest file from the endpoint
+     *
+     * @param errors the errors to add error messages
+     */
+    private void validateEndpoint(final ActionRequestValidationException errors) {
+        if (endpoint == null) {
+            return;
+        }
+
+        try {
+            URL url = new URL(endpoint);
+            url.toURI(); // Validate URL complies with RFC-2396
+            validateManifestFile(url, errors);
+        } catch (MalformedURLException | URISyntaxException e) {
+            log.info("Invalid URL[{}] is provided", endpoint, e);
+            errors.addValidationError("Invalid URL format is provided");
+        }
+    }
+
+    /**
+     * Conduct following validation on url
+     * 1. can read manifest file from the endpoint
+     * 2. the url in the manifest file complies with RFC-2396
+     *
+     * @param url the url to validate
+     * @param errors the errors to add error messages
+     */
+    private void validateManifestFile(final URL url, final ActionRequestValidationException errors) {
+        DatasourceManifest manifest;
+        try {
+            manifest = DatasourceManifest.Builder.build(url);
+        } catch (Exception e) {
+            log.info("Error occurred while reading a file from {}", url, e);
+            errors.addValidationError(String.format(Locale.ROOT, "Error occurred while reading a file from %s: %s", url, e.getMessage()));
```

Comment:
is it safe to include customer provided URL to error message here?

Reply:
Yes. Because this URL is already validated as correct one. Only failed to read file from it.